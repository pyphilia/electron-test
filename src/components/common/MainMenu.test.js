import React from 'react';
import { shallow } from 'enzyme';
import ListItemText from '@material-ui/core/ListItemText';
import MenuItem from '@material-ui/core/MenuItem';
import List from '@material-ui/core/List';
import { Online, Offline } from 'react-detect-offline';
import { MainMenu } from './MainMenu';
import {
  HOME_PATH,
  SPACES_NEARBY_PATH,
  VISIT_PATH,
  LOAD_SPACE_PATH,
  SETTINGS_PATH,
  DEVELOPER_PATH,
  DASHBOARD_PATH,
  SAVED_SPACES_PATH,
} from '../../config/paths';

const MENUITEM_OFFLINE_NUMBER = 2;
const MENUITEM_OFFLINE_ONLINE_COUNT = 11;

const createMainMenuProps = (developerMode, path, authenticated = true) => {
  return {
    t: text => text,
    developerMode,
    history: { replace: jest.fn() },
    match: { path },
    authenticated,
  };
};

describe('<MainMenu />', () => {
  describe('<MainMenu /> with developerMode = false', () => {
    let wrapper;

    beforeAll(() => {
      const props = createMainMenuProps(false, HOME_PATH);
      // eslint-disable-next-line react/jsx-props-no-spreading
      wrapper = shallow(<MainMenu {...props} />);
    });

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('renders one <List /> component', () => {
      expect(wrapper.find(List)).toHaveLength(1);
    });

    it(`renders ${MENUITEM_OFFLINE_NUMBER} online/offline <MenuItem /> components`, () => {
      expect(wrapper.find(Online)).toHaveLength(MENUITEM_OFFLINE_NUMBER);
      expect(wrapper.find(Online).find(MenuItem)).toHaveLength(
        MENUITEM_OFFLINE_NUMBER
      );
      expect(wrapper.find(Offline)).toHaveLength(MENUITEM_OFFLINE_NUMBER);
      expect(wrapper.find(Offline).find(MenuItem)).toHaveLength(
        MENUITEM_OFFLINE_NUMBER
      );
    });

    it(`renders ${MENUITEM_OFFLINE_ONLINE_COUNT} <MenuItem /> components`, () => {
      expect(wrapper.find(MenuItem)).toHaveLength(
        MENUITEM_OFFLINE_ONLINE_COUNT
      );
    });
  });

  describe('<MainMenu /> with developerMode = true', () => {
    let wrapper;

    beforeAll(() => {
      const props = createMainMenuProps(true, HOME_PATH);
      // eslint-disable-next-line react/jsx-props-no-spreading
      wrapper = shallow(<MainMenu {...props} />);
    });

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('renders one <List /> component', () => {
      expect(wrapper.find(List)).toHaveLength(1);
    });

    it(`renders ${MENUITEM_OFFLINE_NUMBER} online/offline <MenuItem /> components`, () => {
      expect(wrapper.find(Online)).toHaveLength(MENUITEM_OFFLINE_NUMBER);
      expect(wrapper.find(Online).find(MenuItem)).toHaveLength(
        MENUITEM_OFFLINE_NUMBER
      );
      expect(wrapper.find(Offline)).toHaveLength(MENUITEM_OFFLINE_NUMBER);
      expect(wrapper.find(Offline).find(MenuItem)).toHaveLength(
        MENUITEM_OFFLINE_NUMBER
      );
    });

    it(`renders ${MENUITEM_OFFLINE_ONLINE_COUNT +
      1} <MenuItem /> components`, () => {
      expect(wrapper.find(MenuItem)).toHaveLength(
        MENUITEM_OFFLINE_ONLINE_COUNT + 1
      );
    });
  });

  describe.each([
    [true, HOME_PATH, 'Home'],
    [true, SAVED_SPACES_PATH, 'Saved Spaces'],
    [true, SPACES_NEARBY_PATH, 'Spaces Nearby'],
    [true, VISIT_PATH, 'Visit a Space'],
    [true, LOAD_SPACE_PATH, 'Load'],
    [true, SETTINGS_PATH, 'Settings'],
    [true, DEVELOPER_PATH, 'Developer'],
    [true, DASHBOARD_PATH, 'Dashboard'],
  ])('<MainMenu /> selects one MenuItem', (developerMode, path, text) => {
    it(`select path=${path} (developerMode = ${developerMode})`, () => {
      const props = createMainMenuProps(developerMode, path);
      // eslint-disable-next-line react/jsx-props-no-spreading
      const wrapper = shallow(<MainMenu {...props} />);
      const selectedMenuItem = wrapper.find({ selected: true });
      expect(selectedMenuItem.length).toBeLessThanOrEqual(2);
      expect(
        selectedMenuItem.map(menuItem =>
          menuItem.find(ListItemText).prop('primary')
        )
      ).toEqual(Array(selectedMenuItem.length).fill(text));
    });
  });
});
