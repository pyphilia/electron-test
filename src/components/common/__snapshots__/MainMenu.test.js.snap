// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainMenu /> <MainMenu /> with developerMode = false renders correctly 1`] = `
<WithStyles(ForwardRef(List))>
  <WithStyles(ForwardRef(MenuItem))
    button={true}
    id="homeMenuItem"
    onClick={[Function]}
    selected={true}
  >
    <WithStyles(ForwardRef(ListItemIcon))>
      <HomeIcon />
    </WithStyles(ForwardRef(ListItemIcon))>
    <WithStyles(ForwardRef(ListItemText))
      primary="Home"
    />
  </WithStyles(ForwardRef(MenuItem))>
  <WithStyles(ForwardRef(MenuItem))
    button={true}
    id="savedSpacesMenuItem"
    onClick={[Function]}
    selected={false}
  >
    <WithStyles(ForwardRef(ListItemIcon))>
      <SaveIcon />
    </WithStyles(ForwardRef(ListItemIcon))>
    <WithStyles(ForwardRef(ListItemText))
      primary="Saved Spaces"
    />
  </WithStyles(ForwardRef(MenuItem))>
  <Offline
    polling={true}
    wrapperType="span"
  >
    <WithStyles(ForwardRef(Tooltip))
      placement="right"
      title="You need an internet connection."
    >
      <div>
        <WithStyles(ForwardRef(MenuItem))
          button={true}
          disabled={true}
          id="spacesNearbyMenuItem"
          onClick={[Function]}
          selected={false}
        >
          <WithStyles(ForwardRef(ListItemIcon))>
            <SearchIcon />
          </WithStyles(ForwardRef(ListItemIcon))>
          <WithStyles(ForwardRef(ListItemText))
            primary="Spaces Nearby"
          />
        </WithStyles(ForwardRef(MenuItem))>
      </div>
    </WithStyles(ForwardRef(Tooltip))>
  </Offline>
  <Online
    polling={true}
    wrapperType="span"
  >
    <WithStyles(ForwardRef(MenuItem))
      button={true}
      id="spacesNearbyMenuItem"
      onClick={[Function]}
      selected={false}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <SearchIcon />
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Spaces Nearby"
      />
    </WithStyles(ForwardRef(MenuItem))>
  </Online>
  <Offline
    polling={true}
    wrapperType="span"
  >
    <WithStyles(ForwardRef(Tooltip))
      placement="right"
      title="You need an internet connection."
    >
      <div>
        <WithStyles(ForwardRef(MenuItem))
          button={true}
          disabled={true}
          id="visitMenuItem"
          onClick={[Function]}
          selected={false}
        >
          <WithStyles(ForwardRef(ListItemIcon))>
            <LanguageIcon />
          </WithStyles(ForwardRef(ListItemIcon))>
          <WithStyles(ForwardRef(ListItemText))
            primary="Visit a Space"
          />
        </WithStyles(ForwardRef(MenuItem))>
      </div>
    </WithStyles(ForwardRef(Tooltip))>
  </Offline>
  <Online
    polling={true}
    wrapperType="span"
  >
    <WithStyles(ForwardRef(MenuItem))
      button={true}
      id="visitMenuItem"
      onClick={[Function]}
      selected={false}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <LanguageIcon />
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Visit a Space"
      />
    </WithStyles(ForwardRef(MenuItem))>
  </Online>
  <WithStyles(ForwardRef(MenuItem))
    button={true}
    id="loadMenuItem"
    onClick={[Function]}
    selected={false}
  >
    <WithStyles(ForwardRef(ListItemIcon))>
      <PublishIcon />
    </WithStyles(ForwardRef(ListItemIcon))>
    <WithStyles(ForwardRef(ListItemText))
      primary="Load"
    />
  </WithStyles(ForwardRef(MenuItem))>
  <WithStyles(ForwardRef(MenuItem))
    button={true}
    id="settingsMenuItem"
    onClick={[Function]}
    selected={false}
  >
    <WithStyles(ForwardRef(ListItemIcon))>
      <SettingsIcon />
    </WithStyles(ForwardRef(ListItemIcon))>
    <WithStyles(ForwardRef(ListItemText))
      primary="Settings"
    />
  </WithStyles(ForwardRef(MenuItem))>
  <WithStyles(ForwardRef(MenuItem))
    button={true}
    id="dashboardMenuItem"
    onClick={[Function]}
    selected={false}
  >
    <WithStyles(ForwardRef(ListItemIcon))>
      <ShowChartIcon />
    </WithStyles(ForwardRef(ListItemIcon))>
    <WithStyles(ForwardRef(ListItemText))
      primary="Dashboard"
    />
  </WithStyles(ForwardRef(MenuItem))>
  <WithStyles(ForwardRef(MenuItem))
    button={true}
    id="signOutMenuItem"
    onClick={[Function]}
    selected={false}
  >
    <WithStyles(ForwardRef(ListItemIcon))>
      <AccountCircleIcon />
    </WithStyles(ForwardRef(ListItemIcon))>
    <WithStyles(ForwardRef(ListItemText))
      primary="Sign Out"
    />
  </WithStyles(ForwardRef(MenuItem))>
  <WithStyles(ForwardRef(MenuItem))
    button={true}
    onClick={[Function]}
  >
    <WithStyles(ForwardRef(ListItemIcon))>
      <ExitToAppIcon />
    </WithStyles(ForwardRef(ListItemIcon))>
    <WithStyles(ForwardRef(ListItemText))
      primary="Quit"
    />
  </WithStyles(ForwardRef(MenuItem))>
</WithStyles(ForwardRef(List))>
`;

exports[`<MainMenu /> <MainMenu /> with developerMode = true renders correctly 1`] = `
<WithStyles(ForwardRef(List))>
  <WithStyles(ForwardRef(MenuItem))
    button={true}
    id="homeMenuItem"
    onClick={[Function]}
    selected={true}
  >
    <WithStyles(ForwardRef(ListItemIcon))>
      <HomeIcon />
    </WithStyles(ForwardRef(ListItemIcon))>
    <WithStyles(ForwardRef(ListItemText))
      primary="Home"
    />
  </WithStyles(ForwardRef(MenuItem))>
  <WithStyles(ForwardRef(MenuItem))
    button={true}
    id="savedSpacesMenuItem"
    onClick={[Function]}
    selected={false}
  >
    <WithStyles(ForwardRef(ListItemIcon))>
      <SaveIcon />
    </WithStyles(ForwardRef(ListItemIcon))>
    <WithStyles(ForwardRef(ListItemText))
      primary="Saved Spaces"
    />
  </WithStyles(ForwardRef(MenuItem))>
  <Offline
    polling={true}
    wrapperType="span"
  >
    <WithStyles(ForwardRef(Tooltip))
      placement="right"
      title="You need an internet connection."
    >
      <div>
        <WithStyles(ForwardRef(MenuItem))
          button={true}
          disabled={true}
          id="spacesNearbyMenuItem"
          onClick={[Function]}
          selected={false}
        >
          <WithStyles(ForwardRef(ListItemIcon))>
            <SearchIcon />
          </WithStyles(ForwardRef(ListItemIcon))>
          <WithStyles(ForwardRef(ListItemText))
            primary="Spaces Nearby"
          />
        </WithStyles(ForwardRef(MenuItem))>
      </div>
    </WithStyles(ForwardRef(Tooltip))>
  </Offline>
  <Online
    polling={true}
    wrapperType="span"
  >
    <WithStyles(ForwardRef(MenuItem))
      button={true}
      id="spacesNearbyMenuItem"
      onClick={[Function]}
      selected={false}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <SearchIcon />
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Spaces Nearby"
      />
    </WithStyles(ForwardRef(MenuItem))>
  </Online>
  <Offline
    polling={true}
    wrapperType="span"
  >
    <WithStyles(ForwardRef(Tooltip))
      placement="right"
      title="You need an internet connection."
    >
      <div>
        <WithStyles(ForwardRef(MenuItem))
          button={true}
          disabled={true}
          id="visitMenuItem"
          onClick={[Function]}
          selected={false}
        >
          <WithStyles(ForwardRef(ListItemIcon))>
            <LanguageIcon />
          </WithStyles(ForwardRef(ListItemIcon))>
          <WithStyles(ForwardRef(ListItemText))
            primary="Visit a Space"
          />
        </WithStyles(ForwardRef(MenuItem))>
      </div>
    </WithStyles(ForwardRef(Tooltip))>
  </Offline>
  <Online
    polling={true}
    wrapperType="span"
  >
    <WithStyles(ForwardRef(MenuItem))
      button={true}
      id="visitMenuItem"
      onClick={[Function]}
      selected={false}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <LanguageIcon />
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Visit a Space"
      />
    </WithStyles(ForwardRef(MenuItem))>
  </Online>
  <WithStyles(ForwardRef(MenuItem))
    button={true}
    id="loadMenuItem"
    onClick={[Function]}
    selected={false}
  >
    <WithStyles(ForwardRef(ListItemIcon))>
      <PublishIcon />
    </WithStyles(ForwardRef(ListItemIcon))>
    <WithStyles(ForwardRef(ListItemText))
      primary="Load"
    />
  </WithStyles(ForwardRef(MenuItem))>
  <WithStyles(ForwardRef(MenuItem))
    button={true}
    id="settingsMenuItem"
    onClick={[Function]}
    selected={false}
  >
    <WithStyles(ForwardRef(ListItemIcon))>
      <SettingsIcon />
    </WithStyles(ForwardRef(ListItemIcon))>
    <WithStyles(ForwardRef(ListItemText))
      primary="Settings"
    />
  </WithStyles(ForwardRef(MenuItem))>
  <WithStyles(ForwardRef(MenuItem))
    button={true}
    id="dashboardMenuItem"
    onClick={[Function]}
    selected={false}
  >
    <WithStyles(ForwardRef(ListItemIcon))>
      <ShowChartIcon />
    </WithStyles(ForwardRef(ListItemIcon))>
    <WithStyles(ForwardRef(ListItemText))
      primary="Dashboard"
    />
  </WithStyles(ForwardRef(MenuItem))>
  <WithStyles(ForwardRef(MenuItem))
    button={true}
    id="developerMenuItem"
    onClick={[Function]}
    selected={false}
  >
    <WithStyles(ForwardRef(ListItemIcon))>
      <CodeIcon />
    </WithStyles(ForwardRef(ListItemIcon))>
    <WithStyles(ForwardRef(ListItemText))
      primary="Developer"
    />
  </WithStyles(ForwardRef(MenuItem))>
  <WithStyles(ForwardRef(MenuItem))
    button={true}
    id="signOutMenuItem"
    onClick={[Function]}
    selected={false}
  >
    <WithStyles(ForwardRef(ListItemIcon))>
      <AccountCircleIcon />
    </WithStyles(ForwardRef(ListItemIcon))>
    <WithStyles(ForwardRef(ListItemText))
      primary="Sign Out"
    />
  </WithStyles(ForwardRef(MenuItem))>
  <WithStyles(ForwardRef(MenuItem))
    button={true}
    onClick={[Function]}
  >
    <WithStyles(ForwardRef(ListItemIcon))>
      <ExitToAppIcon />
    </WithStyles(ForwardRef(ListItemIcon))>
    <WithStyles(ForwardRef(ListItemText))
      primary="Quit"
    />
  </WithStyles(ForwardRef(MenuItem))>
</WithStyles(ForwardRef(List))>
`;
